ws/api/endpoints.go

```diff
@@ -0,0 +1,185 @@
+// Copyright (c) Mainflux
+// SPDX-Licence-Identifier: Apache-2.0
+
+package api
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"net/url"
+	"regexp"
+	"strings"
+
+	"github.com/go-zoo/bone"
+	"github.com/gorilla/websocket"
+	"github.com/mainflux/mainflux/pkg/errors"
+	"github.com/mainflux/mainflux/ws"
+)
+
+var channelPartRegExp = regexp.MustCompile(`^/channels/([\w\-]+)/messages(/[^?]*)?(\?.*)?$`)
+
+func handshake(svc ws.Service) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		req, err := decodeRequest(r)
+		if err != nil {
+			switch err {
+			case ws.ErrEmptyID:
+				w.WriteHeader(http.StatusServiceUnavailable)
+				return
+			case errUnauthorizedAccess:
+				w.WriteHeader(http.StatusForbidden)
+				return
+			case errMalformedSubtopic:
+				w.WriteHeader(http.StatusBadRequest)
+				return
+			case errors.ErrMalformedEntity:
+				w.WriteHeader(http.StatusBadRequest)
+				return
+			default:
+				logger.Warn(fmt.Sprintf("Failed to authorize: %s", err.Error()))
+				w.WriteHeader(http.StatusServiceUnavailable)
+				return
+			}
+		}
+
+		// Upgrade to a new ws connection.
+		conn, err := upgrader.Upgrade(w, r, nil)
+		if err != nil {
+			logger.Warn(fmt.Sprintf("Failed to upgrade connection to websocket: %s", err.Error()))
+			return
+		}
+
+		req.conn = conn
+		client := ws.NewClient(conn, "")
+
+		// Subscribe using the adapterservice
+		if err := svc.Subscribe(context.Background(), req.thingKey, req.chanID, req.subtopic, client); err != nil {
+			logger.Warn(fmt.Sprintf("Failed to subscribe to broker: %s", err.Error()))
+			switch err {
+			case ws.ErrEmptyID, ws.ErrEmptyTopic:
+				w.WriteHeader(http.StatusBadRequest)
+			case ws.ErrInvalidConnection:
+				w.WriteHeader(http.StatusUnauthorized)
+				req.conn.Close()
+				return
+			case ws.ErrAlreadySubscribed:
+				req.conn.Close()
+				return
+			case ws.ErrFailedConnection:
+				w.WriteHeader(http.StatusServiceUnavailable)
+				req.conn.Close()
+				return
+			default:
+				w.WriteHeader(http.StatusUnauthorized)
+				req.conn.Close()
+				return
+			}
```

Comment:
Extract this to `encodeError`.

Reply:
Please take a look at the `encodeErro()` function, 
 I will switch to your implementation of `encodeError()`, if current one isn't correct